### ISSUE TITLE
JavaParserInterfaceDeclaration.toReferenceType Infinite Loop

### ISSUE THREAD

**Issue by FRuffy** [javaparser/javasymbolsolver#185]
It can happen that toReferenceType in this class loops infinitely.
Example:
`IntStream extends BaseStream<Integer, IntStream>`
tries to resolve Integer and IntStream, which causes a cycle as IntStream resolves to BaseStream and to IntStream again. 

I temporarly added a filter which removes nodes that are actually the wrappedNode itself to mitigate the problem

```
List<com.github.javaparser.ast.type.Type> superClassTypeParameters = classOrInterfaceType.getTypeArguments().get();

for (com.github.javaparser.ast.type.Type type : superClassTypeParameters) {
	if (type.toString().equals(this.getWrappedNode().getNameAsString())) {
		superClassTypeParameters.remove(type);
	}
}

List<Type> solvedTypeParameters = superClassTypeParameters.stream()
.map(ta -> JavaParserFacade.get(typeSolver).convert(ta, ta)).collect(Collectors.toList());

return new ReferenceTypeImpl(ref.getCorrespondingDeclaration().asReferenceType(), solvedTypeParameters,typeSolver);
```

However, it is less than ideal and quite hideous. Unfortunately I could not get a stream filter solution to work. Is there maybe a better solution?

**Comment by FRuffy** [comment]
Okay let me see..
```import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;

import com.github.javaparser.JavaParser;
import com.github.javaparser.ParseException;
import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.expr.MethodCallExpr;
import com.github.javaparser.ast.stmt.ReturnStmt;
import com.github.javaparser.ast.visitor.VoidVisitorAdapter;
import com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade;
import com.github.javaparser.symbolsolver.model.declarations.MethodDeclaration;
import com.github.javaparser.symbolsolver.model.resolution.SymbolReference;
import com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver;
import com.github.javaparser.symbolsolver.resolution.typesolvers.JavaParserTypeSolver;

public class PrintExpressionType {

	static class TypeCalculatorVisitor extends VoidVisitorAdapter<JavaParserFacade> {
		@Override
		public void visit(ReturnStmt n, JavaParserFacade javaParserFacade) {
			super.visit(n, javaParserFacade);
			// System.out.println(n.getExpr().toString() + " has type " +
			// javaParserFacade.getType(n.getExpr()));
		}

		@Override
		public void visit(MethodCallExpr n, JavaParserFacade javaParserFacade) {
			System.out.println("Solving... "+n.toString());
			SymbolReference<MethodDeclaration> methodRef = null;
			try {
				methodRef = javaParserFacade.solve(n);
			} catch (RuntimeException e) {
				System.err.println("ERROR: Could not solve method call: " + e.getMessage());
				return;
			}
			if (methodRef.isSolved()) {
				MethodDeclaration methodDecl = methodRef.getCorrespondingDeclaration();
				System.out.println(methodDecl.getQualifiedName()+ " has type: "+ methodDecl.getReturnType().describe() );
			} else {
				System.out.println("Warning: Could not solve method call: " + n.getNameAsString());
			}
		}
	}

	public static void main(String[] args) throws FileNotFoundException, ParseException {
		System.out.println("Working Directory = " + System.getProperty("user.dir"));
		CombinedTypeSolver typeSolver = new CombinedTypeSolver();

		typeSolver.add(new JavaParserTypeSolver(new File("../../jre_library/")));
		CompilationUnit agendaCu = JavaParser.parse(new FileInputStream(new File("./util/Arrays.java")));

		agendaCu.accept(new TypeCalculatorVisitor(), JavaParserFacade.get(typeSolver));

	}

}
```
This is a minimal working example of the approach I have.
"jre_library" is the src.zip of the jdk with all the raw java files in it.

Arrays.java is the java file of java.util. I am trying to "solve" all method calls in these method declarations. In general, it works quite well but I am frequently running into problems. I guess parsing the entire Java library is a good way to test the tool lol.
I used the latest master version of the git repo, I attached the corresponding output of a test run.

[bugs.txt](https://github.com/javaparser/javasymbolsolver/files/870317/bugs.txt)

**Comment by xdrop** [comment]
I've created a (failing) test which reproduces this (0c7b22e96e55be5c762c430aabd10e6da9364165). Not sure how to fix this yet however, the filtering solution suggested above won't work for all cases . 

We'd need some way of checking whether we are stuck in the recursion chain

**Comment by loren138** [comment]
I get stuck in an infinate loop trying to solve this function:
https://github.com/tharindum/opennms_dashboard/blob/master/features/system-report/src/main/java/org/opennms/systemreport/opennms/OpenNMSReportPlugin.java#L101

  38674 OnmsIpInterface
  38675 ctu OnmsIpInterface OnmsIpInterface
  38676 solveType org.opennms.netmgt.dao.IpInterfaceDao
  38677 OnmsDao<OnmsIpInterface, Integer> org.opennms.netmgt.dao.IpInterfaceDao
  38678 OnmsIpInterface
  38679 ctu OnmsIpInterface OnmsIpInterface
  38680 solveType org.opennms.netmgt.dao.IpInterfaceDao
  38681 OnmsDao<OnmsIpInterface, Integer> org.opennms.netmgt.dao.IpInterfaceDao
  38682 OnmsIpInterface
  38683 ctu OnmsIpInterface OnmsIpInterface
  38684 solveType org.opennms.netmgt.dao.IpInterfaceDao
  38685 OnmsDao<OnmsIpInterface, Integer> org.opennms.netmgt.dao.IpInterfaceDao

Somehow the context keeps being ``org.opennms.netmgt.dao.IpInterfaceDao`` (``this.typeDeclaration`` in ``JavaParserTypeDeclarationAdapter``).  The best answer I've come up with so far is to have a static list tracking what I've looped into already to stop me from going into it again.

In my experiment, solving with a hashmap allowing 10 recursions solved 486 methods that I was tracking, but took over 5 hours.  The original code solved, 391 methods in 50 minutes.  Using a linked list (so allowing no recursion) solved 481 methods in 23 minutes.

For the linked list, I added ``private static List<String> inLoop = new LinkedList<>();`` to the ``JavaParserTypeDeclarationAdapter`` class and changed the function to look like the below.  The implementation using a Hashmap<String, Integer> is in comments.  Ideally, I think you would pass recursion as a parameter through the whole loop, but I didn't want to make that many modifications.

```java
    public SymbolReference<TypeDeclaration> solveType(String name, TypeSolver typeSolver) {
        if (this.wrappedNode.getName().getId().equals(name)) {
            return SymbolReference.solved(JavaParserFacade.get(typeSolver).getTypeDeclaration(wrappedNode));
        }

        // Internal classes
        for (BodyDeclaration<?> member : this.wrappedNode.getMembers()) {
            if (member instanceof com.github.javaparser.ast.body.TypeDeclaration) {
                com.github.javaparser.ast.body.TypeDeclaration<?> internalType = (com.github.javaparser.ast.body.TypeDeclaration<?>) member;
                if (internalType.getName().getId().equals(name)) {
                    return SymbolReference.solved(JavaParserFacade.get(typeSolver).getTypeDeclaration(internalType));
                } else if (name.startsWith(String.format("%s.%s", wrappedNode.getName(), internalType.getName()))) {
                    return JavaParserFactory.getContext(internalType, typeSolver).solveType(name.substring(wrappedNode.getName().getId().length() + 1), typeSolver);
                } else if (name.startsWith(String.format("%s.", internalType.getName()))) {
                    return JavaParserFactory.getContext(internalType, typeSolver).solveType(name.substring(internalType.getName().getId().length() + 1), typeSolver);
                }
            }
        }

        if (wrappedNode instanceof NodeWithTypeParameters) {
            NodeWithTypeParameters<?> nodeWithTypeParameters = (NodeWithTypeParameters<?>) wrappedNode;
            for (TypeParameter astTpRaw : nodeWithTypeParameters.getTypeParameters()) {
                TypeParameter astTp = astTpRaw;
                if (astTp.getName().getId().equals(name)) {
                    return SymbolReference.solved(new JavaParserTypeParameter(astTp, typeSolver));
                }
            }
        }

        //System.out.println("solveType "+this.typeDeclaration.getQualifiedName());

        // Look into extended classes and implemented interfaces
        //if (!inLoop.containsKey(this.typeDeclaration.getQualifiedName())) {
        //   inLoop.put(this.typeDeclaration.getQualifiedName(), 0);
        //}
        //Integer value = inLoop.get(this.typeDeclaration.getQualifiedName());
        //if (value < 10) {
        if (!inLoop.contains(this.typeDeclaration.getQualifiedName())) {
            inLoop.add(this.typeDeclaration.getQualifiedName());
            //inLoop.replace(this.typeDeclaration.getQualifiedName(), value + 1);
            for (ReferenceType ancestor : this.typeDeclaration.getAncestors()) {
                try {
                    for (TypeDeclaration internalTypeDeclaration : ancestor.getTypeDeclaration().internalTypes()) {
                        if (internalTypeDeclaration.getName().equals(name)) {
                            return SymbolReference.solved(internalTypeDeclaration);
                        }
                    }
                } catch (UnsupportedOperationException e) {
                    // just continue using the next ancestor
                }

            }
            //inLoop.replace(this.typeDeclaration.getQualifiedName(), inLoop.get(this.typeDeclaration.getQualifiedName()) - 1);
            inLoop.remove(this.typeDeclaration.getQualifiedName());
        //} else {
            //System.out.println("skip loop");
        }

        return context.getParent().solveType(name, typeSolver);
    }

```

**Comment by staslev** [comment]
I've been bitten by this as well.

If you are happy with your solution, would you be willing to submit a PR?

**Comment by ftomassetti** [comment]
Importing the test written by @xdrop the problem seems fixed

**Comment by ftomassetti** [comment]
I think it was fixed by @Gottox when fixing #232 by using LazyType . I am going to close this now but feel free to reopen it if the problem persists
